plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs

```diff
@@ -1207,301 +1313,495 @@ instance uni ~ DefaultUni => ToBuiltinMeaning uni DefaultFun where
+
     -- Strings
     toBuiltinMeaning _semvar AppendString =
-        makeBuiltinMeaning
-            ((<>) @Text)
+        let appendStringDenotation :: Text -> Text -> Text
+            appendStringDenotation = (<>)
+            {-# INLINE appendStringDenotation #-}
+        in makeBuiltinMeaning
+            appendStringDenotation
             (runCostingFunTwoArguments . paramAppendString)
+
     toBuiltinMeaning _semvar EqualsString =
-        makeBuiltinMeaning
-            ((==) @Text)
+        let equalsStringDenotation :: Text -> Text -> Bool
+            equalsStringDenotation = (==)
+            {-# INLINE equalsStringDenotation #-}
+        in makeBuiltinMeaning
+            equalsStringDenotation
             (runCostingFunTwoArguments . paramEqualsString)
+
     toBuiltinMeaning _semvar EncodeUtf8 =
-        makeBuiltinMeaning
-            encodeUtf8
+        let encodeUtf8Denotation :: Text -> BS.ByteString
+            encodeUtf8Denotation = encodeUtf8
+            {-# INLINE encodeUtf8Denotation #-}
+        in makeBuiltinMeaning
+            encodeUtf8Denotation
             (runCostingFunOneArgument . paramEncodeUtf8)
+
     toBuiltinMeaning _semvar DecodeUtf8 =
-        makeBuiltinMeaning
-            (reoption @_ @EvaluationResult . decodeUtf8')
+        let decodeUtf8Denotation :: BS.ByteString -> EvaluationResult Text
+            decodeUtf8Denotation = reoption . decodeUtf8'
+            {-# INLINE decodeUtf8Denotation #-}
+        in makeBuiltinMeaning
+            decodeUtf8Denotation
             (runCostingFunOneArgument . paramDecodeUtf8)
+
     -- Bool
     toBuiltinMeaning _semvar IfThenElse =
-        makeBuiltinMeaning
-            (\b x y -> if b then x else y)
+        let ifThenElseDenotation :: Bool -> a -> a -> a
+            ifThenElseDenotation b x y = if b then x else y
+            {-# INLINE ifThenElseDenotation #-}
+        in makeBuiltinMeaning
+            ifThenElseDenotation
             (runCostingFunThreeArguments . paramIfThenElse)
+
     -- Unit
     toBuiltinMeaning _semvar ChooseUnit =
-        makeBuiltinMeaning
-            (\() a -> a)
+        let chooseUnitDenotation :: () -> a -> a
+            chooseUnitDenotation () x = x
+            {-# INLINE chooseUnitDenotation #-}
+        in makeBuiltinMeaning
+            chooseUnitDenotation
             (runCostingFunTwoArguments . paramChooseUnit)
+
     -- Tracing
     toBuiltinMeaning _semvar Trace =
-        makeBuiltinMeaning
-            (\text a -> a <$ emit text)
+        let traceDenotation :: Text -> a -> Emitter a
+            traceDenotation text a = a <$ emit text
+            {-# INLINE traceDenotation #-}
+        in makeBuiltinMeaning
+            traceDenotation
             (runCostingFunTwoArguments . paramTrace)
+
     -- Pairs
     toBuiltinMeaning _semvar FstPair =
-        makeBuiltinMeaning
-            fstPlc
+        let fstPairDenotation :: SomeConstant uni (a, b) -> EvaluationResult (Opaque val a)
```

Comment:
I should read the documentation again, but why does this need `EvaluationResult`?  I initially thought it was because it'll fail if you give it a non-pair, but anything can fail if you give it something of the wrong type.  Does this have to do with #-polymorphism, like it could in theory get a pair with an entry that isn't a built-in constant?

Reply:
>  I initially thought it was because it'll fail if you give it a non-pair, but anything can fail if you give it something of the wrong type. Does this have to do with #-polymorphism, like it could in theory get a pair with an entry that isn't a built-in constant?

No, a pair always has constants inside at runtime, there's no way one could sneak anything else in there, since it's unrepresentable in the AST.

This is described in the docs in this file but tl;dr is that we cannot have 

```haskell
fstPairDenotation :: (a, b) -> Opaque val a
```

due to some technical and annoying reasons, i.e. we cannot automatically unlift values of unmonomorphized polymorphic built-in types in the builtins machinery, unlike values of monomorphic types (e.g. `Integer`) or values of monomorphized polymorphic built-in types (e.g. `[(Bool, ByteString)]`). Therefore we have to unlift a pair manually within the denotation of the builtin and that operation can fail, hence `EvaluationResult`.

_In theory_ we could make automatic unlifting work for `fstPair`. But that is non-trivial and making automatic unlifting work for something like `mkCons` sounds like a complete nightmare, since we need to check equality of type tags extracted from the two arguments there and doing that automatically is just way too complicated and I'm sure nobody would appreciate the complexity of such a machinery, hence I didn't attempt to implement that. Present-day situation is "if the builtin doesn't have unmonomorphized polymorphic built-in types, the machinery does everything for you, but otherwise you have to understand how to do things manually" and I think it's a fine balance between complexity of the machinery and it's user-friendliness.